src/actions/pkg.cr

```diff
@@ -1,147 +1,129 @@
+        return {data: result} if result.any?
+        bug = InternalServerError.new context, "received empty set from Prefix#clean_unused_packages; please report this strange bug"
+        error = NoPkgsToClean.new context
+        error.cause = bug
+        raise error
+      end
+      raise NoPkgsToClean.new context
     end
-    raise NoPkgsToClean.new context
-  end
 
-  # List built packages
-  relative_get nil do |context|
-    if context.current_user? && Actions.has_access? context, Access::Read
-      pfx = DPPM::Prefix.new prefix
-      # Build JSON directly to the HTTP response IO
-      JSON.build context.response do |json|
-        json.array do
-          # build an array of all the package names
-          pfx.each_pkg { |pkg| json.string pkg.package }
+    # List built packages
+    relative_get nil do |context|
+      if context.current_user? && Actions.has_access? context, Access::Read
+        # Build JSON directly to the HTTP response IO
+        build_json context.response do |json|
+          json.array do
+            # build an array of all the package names
+            get_prefix_or_default(from: context).each_pkg { |pkg| json.string pkg.package }
+          end
         end
+        next context
       end
-      next context
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Clean unused built packages
-  relative_delete "/clean" do |context|
-    if context.current_user? && Actions.has_access? context, Access::Delete
-      clean_unused_packages(context).to_json context.response
-      context.response.flush
-      next context
+    # Clean unused built packages
+    relative_delete "/clean" do |context|
+      if context.current_user? && Actions.has_access? context, Access::Delete
+        clean_unused_packages(context).to_json context.response
+        context.response.flush
+        next context
+      end
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Query information about a given package
-  relative_get "/:id/query" do |context|
-    if context.current_user? && Actions.has_access? context, Access::Read
-      package_name = URI.unescape context.params.url["id"]
-      # Iterate over the packages to find the relevant one.
-      if selected_pkg = find_package_by_name package_name, prefix
-        # The package whose name was the :id URL parameter was found and we can query it
+    # Query information about a given package
+    relative_get "/:id/query" do |context|
+      if context.current_user? && Actions.has_access? context, Access::Read
+        package_name = URI.unescape context.params.url["id"]
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = get_prefix_or_default(from: context).new_pkg package_name, version
+        raise NoSuchPackage.new context, package_name if selected_pkg.nil?
```

Comment:
Oh, `new_pkg` and `Pkg.create` don't return `Pkg?` so how will I know if the package hasn't been built?

Reply:
It does. `Pkg` is package object, a virtual package if you prefer.
I may create `VirtualPkg` and `Pkg`, not done yet. For now, you can use `#exists?` and `#exists!` (`program_data.cr`)